Kernal context can be left open forever if fail to generate taichi AST.

**To Reproduce**

```py
import taichi
vars = taichi.field(dtype=taichi.uint16, shape=(256,))

@taichi.kernel
def name_error():
    for v in vars:
        v += offset  # It raises a NameError as expected

try:
    name_error()
except:
    pass

# AssertionError: field cannot be called in Taichi-scope
vars2 = taichi.field(dtype=taichi.uint16, shape=(256,))
```

**Potential Fix**

```diff
            self.runtime.inside_kernel = True
            self.runtime.current_kernel = self
            compiled()
-           self.runtime.inside_kernel = False
-           self.runtime.current_kernel = None
+           try:
+               compiled()
+           finally:
+               self.runtime.inside_kernel = False
+               self.runtime.current_kernel = None

        taichi_kernel = taichi_kernel.define(taichi_ast_generator)
        self.kernel_cpp = taichi_kernel
```

**Additional comments**

I understand that the snippet above is weird since nobody would code like that. However, I am working on a project to integrate Taichi with some popular software like UE4. The project exposes a multi-line string editor to the software's GUI and performs a serial Python import and reload, which re-compiles the Taichi kernels.

Obviously, if the user inputs a typo, Taichi will be disabled unless I manually reset the `PyTaichi`. I've posted my fix, and I'm not sure if it's a decent solution. Please let me know if it would be accepted, and maybe there are better ideas according to my use case.

--- a/python/taichi/lang/kernel_impl.py
+++ b/python/taichi/lang/kernel_impl.py
@@ -443,9 +443,11 @@ class Kernel:
                 )
             self.runtime.inside_kernel = True
             self.runtime.current_kernel = self
-            compiled()
-            self.runtime.inside_kernel = False
-            self.runtime.current_kernel = None
+            try:
+                compiled()
+            finally:
+                self.runtime.inside_kernel = False
+                self.runtime.current_kernel = None
 
         taichi_kernel = taichi_kernel.define(taichi_ast_generator)
         self.kernel_cpp = taichi_kernel

